### PR TITLE
add events to IntVoteInterface, avatar to those events

### PR DESCRIPTION
**Pull request by dkent600**
Resolves: #450 
Resolves: #452 

I'm uncertain about one aspect of this, which is the mismatch in events between `IntVoteInterface` and `GenesisProtocol` and `AbsoluteVote`:

- GP doesn't declare the `CancelProposal` and `CancelVoting` events
- GP `ExecuteProposal` has an extra param: `ExecutionState _executionState`
- AV `VoteProposal` has an extra param: `bool isOwnerVote` param

Maybe this is all not a problem, but any suggestions are invited about how to better address these differences.

[Edit:]  Decision for above was to add new `AVVoteProposal` and `GPExecuteProposal` events.

### CODE REVIEW

File: contracts/VotingMachines/AbsoluteVote.sol [hunk 1]
```diff
@@ -95,11 +95,13 @@ contract AbsoluteVote is IntVoteInterface {
    * @param _proposalId the proposal ID
    */
     function cancelProposal(bytes32 _proposalId) external onlyProposalOwner(_proposalId) votable(_proposalId) returns(bool) {
+
         if (! parameters[proposals[_proposalId].paramsHash].allowOwner) {
             return false;
         }
+        address avatar = proposals[_proposalId].avatar;
```
Comment (orenyodfat) [inline contracts/VotingMachines/AbsoluteVote.sol]: how about emit the event before deleting the proposal and save this assignment ?

File: contracts/VotingMachines/IntVoteInterface.sol [hunk 2]
```diff
@@ -9,6 +9,12 @@ interface IntVoteInterface {
     modifier onlyProposalOwner(bytes32 _proposalId) {revert(); _;}
     modifier votable(bytes32 _proposalId) {revert(); _;}
 
+    event NewProposal(bytes32 indexed _proposalId, address indexed _avatar, uint _numOfChoices, address _proposer, bytes32 _paramsHash);
```
Comment (orenyodfat) [inline contracts/VotingMachines/IntVoteInterface.sol]: please remove this from the interface.

File: contracts/VotingMachines/AbsoluteVote.sol (hunk 1 above)
Comment (dkent600) [reply contracts/VotingMachines/AbsoluteVote.sol]: Are you sure? The pattern in all the other contracts is to emit *after* deletion.

File: contracts/VotingMachines/IntVoteInterface.sol (hunk 2 above)
Comment (dkent600) [reply contracts/VotingMachines/IntVoteInterface.sol]: Can you please explain why, for the record?

Comment (orenyodfat) [reply contracts/VotingMachines/IntVoteInterface.sol]: if it is the intrerface so remove that from the GenesisProtocol and AbsoluteVote

Comment (dkent600) [reply contracts/VotingMachines/IntVoteInterface.sol]: Done

Comment (dkent600) [reply contracts/VotingMachines/IntVoteInterface.sol]: AbsoluteVote adds a parameter, so I left it in.  Same deal for `GP.ExecuteProposal`.  (See my comments in the PR description.)

Comment (orenyodfat) [reply contracts/VotingMachines/IntVoteInterface.sol]: so remove it from the interface

Comment (orenyodfat) [reply contracts/VotingMachines/IntVoteInterface.sol]: To many unneeded events.

Comment (dkent600) [reply contracts/VotingMachines/IntVoteInterface.sol]: OK, then I would vote to eliminate `AbsoluteVoteProposal` (the existing event in AV) if it is unneeded.  The `IntVoteProposal` version *is* needed.

Similarly with `ExecuteProposal` in GP.

We need to be able to subscribe to these events without knowing what the voting machine is.  You OK with that?

Comment (orenyodfat) [reply contracts/VotingMachines/IntVoteInterface.sol]: AbsoluteVoteProposal does not exist in ARC.
you mean VoteProposal ?

Comment (dkent600) [reply contracts/VotingMachines/IntVoteInterface.sol]: I describe what I mean by "AbsoluteVoteProposal" in my comment in this thread just above.

Comment (dkent600) [reply contracts/VotingMachines/IntVoteInterface.sol]: But yes, I would propose to eliminate the existing `AbsoluteVote.VoteProposal` and `GenesisProtocol.ExecuteProposal` events if you don't like my proposal above to rename and emit them as separate events.

Comment (dkent600) [reply contracts/VotingMachines/IntVoteInterface.sol]: @orenyodfat 

This is a long and confusing thread, and it needs to be resolved, so let me summarize where we are at on it:

There is a mismatch in two different events between IntVoteInterface and GenesisProtocol and AbsoluteVote:

* GP.ExecuteProposal has an extra param: ExecutionState _executionState
* AV.VoteProposal has an extra param: bool isOwnerVote param

In order to conform to the IntVoteInterface, AV and GP needs to emit the exact events as they are defined in the interface.  As far as I know, there are two ways to do this:

1) Remove the extra parameters from the extended events in AV and GP
2) Have AV and GP throw both the interface version and extended versions of the events.  We would rename the extended events.

Can you please indicate which is the preferred option, thanks?

File: contracts/VotingMachines/AbsoluteVote.sol [hunk 3]
```diff
@@ -98,11 +94,13 @@ contract AbsoluteVote is IntVoteInterface {
    * @param _proposalId the proposal ID
    */
     function cancelProposal(bytes32 _proposalId) external onlyProposalOwner(_proposalId) votable(_proposalId) returns(bool) {
+
```
[the remark this answers is not in the document]
Comment (dkent600) [reply contracts/VotingMachines/AbsoluteVote.sol]: Removed

File: contracts/VotingMachines/IntVoteInterface.sol (hunk 2 above)
Comment (dkent600) [reply contracts/VotingMachines/IntVoteInterface.sol]: Decision is to add new events: AVVoteProposal and GPExecuteProposal.  This is done.

File: contracts/VotingMachines/AbsoluteVote.sol [hunk 4]
```diff
@@ -31,7 +31,7 @@ contract AbsoluteVote is IntVoteInterface {
         bool open; // voting open flag
     }
 
-    event VoteProposal(bytes32 indexed _proposalId, address indexed _avatar, address indexed _voter, uint _vote, uint _reputation, bool _isOwnerVote);
+    event AVVoteProposal(bytes32 indexed _proposalId, address indexed _avatar, address indexed _voter, uint _vote, uint _reputation, bool _isOwnerVote);
```
Comment (orenyodfat) [inline contracts/VotingMachines/AbsoluteVote.sol]: the AVVoteProposal should include only the extended params which are not included on VoteProposal interface (with the proposalId as Index).

File: contracts/VotingMachines/GenesisProtocol.sol [hunk 5]
```diff
@@ -80,7 +80,7 @@ contract GenesisProtocol is IntVoteInterface,UniversalScheme {
         mapping(address=>Staker) stakers;
     }
 
-    event ExecuteProposal(bytes32 indexed _proposalId,
+    event GPExecuteProposal(bytes32 indexed _proposalId,
```
Comment (orenyodfat) [inline contracts/VotingMachines/GenesisProtocol.sol]: the GPExecuteProposal should include only the extended params which are not included on ExecuteProposal common interface (with the proposalId as Index).

File: contracts/VotingMachines/AbsoluteVote.sol [hunk 6]
```diff
@@ -31,12 +31,8 @@ contract AbsoluteVote is IntVoteInterface {
         bool open; // voting open flag
     }
 
-    event NewProposal(bytes32 indexed _proposalId, uint _numOfChoices, address _proposer, bytes32 _paramsHash);
-    event CancelProposal(bytes32 indexed _proposalId);
-    event ExecuteProposal(bytes32 indexed _proposalId, uint _decision, uint _totalReputation);
-    event VoteProposal(bytes32 indexed _proposalId, address indexed _voter, uint _vote, uint _reputation, bool _isOwnerVote);
-    event CancelVoting(bytes32 indexed _proposalId, address indexed _voter);
-    event RefreshReputation(bytes32 indexed _proposalId, address indexed _voter,uint _reputation);
+    event AVVoteProposal(bytes32 indexed _proposalId, address indexed _avatar, address indexed _voter, uint _vote, uint _reputation, bool _isOwnerVote);
```
Comment (orenyodfat) [inline contracts/VotingMachines/AbsoluteVote.sol]: AVVoteProposal should include only the extended param which is not included at the interface event (plus the proposalId as Index)..
There is no need to duplicate .

Comment (dkent600) [reply contracts/VotingMachines/AbsoluteVote.sol]: Done

File: contracts/VotingMachines/GenesisProtocol.sol (hunk 5 above)
Comment (dkent600) [reply contracts/VotingMachines/GenesisProtocol.sol]: Done

File: contracts/VotingMachines/AbsoluteVote.sol (hunk 6 above)
Comment (orenyodfat) [reply contracts/VotingMachines/AbsoluteVote.sol]: Why the _avatar  and _voter are needed indexes at the AVVoteProposal ?
One can fetch the proposalId for VoteProposal (by _avatar or _voter) and then look for AVVoteProposal according to _proposalId .
These two events are emitted one after the other.

File: contracts/VotingMachines/GenesisProtocol.sol (hunk 5 above)
Comment (orenyodfat) [reply contracts/VotingMachines/GenesisProtocol.sol]: Does the index  _avatar is really needed here ?
One can fetch the _proposalId via the _avatar from ExecuteProposal .

Comment (dkent600) [reply contracts/VotingMachines/GenesisProtocol.sol]: > Does the index _avatar is really needed here?

I left it in to maintain compatibility with all of the other GP events, as well as all of the events in AbsoluteVote and IntVoteInterrface and almost every other scheme event in Arc.

The only inconsistencies are a handful of events in VestingScheme (five) and VoteInOrganizationScheme (one).  I think we should seriously consider adding indexed _avatar to those as well.

This kind of API consistency is very important in supporting consumers of any library or framework, including this one.

File: contracts/VotingMachines/AbsoluteVote.sol (hunk 6 above)
Comment (dkent600) [reply contracts/VotingMachines/AbsoluteVote.sol]: @orenyodfat  This is for API consistency.  Is there a cost to having this indexed _avatar here?

File: contracts/VotingMachines/AbsoluteVote.sol (hunk 4 above)
Comment (dkent600) [reply contracts/VotingMachines/AbsoluteVote.sol]: Removed

File: contracts/VotingMachines/GenesisProtocol.sol (hunk 5 above)
Comment (dkent600) [reply contracts/VotingMachines/GenesisProtocol.sol]: Removed

**Pull request merged**